CORE-6014 - Allow real JSON to be supplied as FlowStartParams via JsonObject request type

conal-smith opened the pull request:
Real JSON can be supplied in start flow params and JsonObject can be used in any other request type to allow real JSON object to be sent rather than escaping JSON.
- Add support for marshalling and unmarshalling `JsonObject` in configured Javalin Jackson object mapper
- Add JsonObject to StartFlowParameters.kt so escaped JSON or real JSON can be provided as part of payload.
- test arrays, objects, maps, values.
- Also supports escaped JSON as before.
- Response objects that return JsonObject return escaped JSON rather than JSON object.

Working examples:
```
POST https://localhost:8888/api/v1/flow/6D5296318533
{
    "clientRequestId": "launchpad-7",
    "flowClassName": "net.cordapp.flowworker.development.flows.RpcSmokeTestFlow",
    "requestData": "{\"command\":\"echo\", \"data\":{\"echo_value\": \"helloconal\"}}"
}
```
```
POST https://localhost:8888/api/v1/flow/6D5296318533
{
    "clientRequestId": "launchpad-6",
    "flowClassName": "net.cordapp.flowworker.development.flows.RpcSmokeTestFlow",
    "requestData": {
      "command":"echo", 
      "data":{
        "echo_value": "helloconal"
      }
    }
}
```

conal-smith committed:
CORE-6014 - add JsonObject and support for marshalling and unmarshalling the type to String in configured Javalin Jackson object mapper
- add JsonObject to StartFlowParameters.kt so escaped JSON or real JSON can be provided as part of payload.
- add jackson serializer and deserializer for specific types of nodes.
- test arrays, objects, maps, values.

conal-smith committed:
CORE-5679 update swagger baseline with new json object type

sprintengine commented on components/flow/flow-rpcops-service-impl/src/main/kotlin/net/corda/flow/rpcops/impl/v1/FlowRPCOpsImpl.kt:
```diff
@@ -89,7 +89,7 @@ class FlowRPCOpsImpl @Activate constructor(
                 vNode,
                 flowClassName,
-                startFlow.requestData,
+                startFlow.requestData.toString(),
```
We are sending the same escaped string to the flow. Marshaller builds a JsonObjectAsString with value obtained by the textValue / toString of the TreeNode.

sprintengine commented on libs/http-rpc/http-rpc-server-impl/src/main/kotlin/net/corda/httprpc/server/impl/internal/HttpRPCServerSerialization.kt:
```diff
@@ -26,3 +36,25 @@ internal object MemberX500NameSerializer : JsonSerializer<MemberX500Name>() {
+        val str = if (treeNode.isObject) {
+            (treeNode as ObjectNode).toString()
+        } else if (treeNode.isValueNode) {
+            (treeNode as ValueNode).textValue()
```
toString here would result in double escaping

sprintengine commented on libs/http-rpc/http-rpc-server-impl/src/main/kotlin/net/corda/httprpc/server/impl/internal/HttpRPCServerSerialization.kt:
```diff
@@ -26,3 +36,25 @@ internal object MemberX500NameSerializer : JsonSerializer<MemberX500Name>() {
+
+internal object JsonObjectSerializer : JsonSerializer<JsonObject>() {
+    override fun serialize(obj: JsonObject, generator: JsonGenerator, provider: SerializerProvider) {
+        generator.writeString(obj.toString())
```
If an RPCOps method returns a JsonObject this serializer will just writes the string representation of the object i.e. escaped string:
```
{"id":"aaa123","obj":"{\"message\":\"Hey Mars\",\"planetaryOnly\":\"true\",\"target\":\"C=GB, L=FOURTH, O=MARS, OU=PLANET\"}"}
```

It might be possible to write out the actual json object and omit JsonObjectAsString's "value" field.

sprintengine commented on components/flow/flow-rpcops-service/src/main/kotlin/net/corda/flow/rpcops/v1/types/request/StartFlowParameters.kt:
```diff
@@ -10,5 +12,5 @@ package net.corda.flow.rpcops.v1.types.request
     val clientRequestId: String,
     val flowClassName: String,
-    val requestData: String
+    val requestData: JsonObject
```
We could also use this for `UpdateConfigParameters.config` and `GetConfigResponse.configWithDefaults` and possibly `sourceConfig` and I'm sure there are other places.

lankydan replied:
Just to clarify, however I think I know the answer, `JsonObject` can be a `String`, `Int` or an `Array` as well right? Flows _might_ want to still receive just a string so non json data can be received, e.g. xml or csv.

sprintengine replied:
Yeah if the treeNode is just a string (ValueNode) it just gets textValue() and sets into JsonObjectAsString, FlowRpcOpsImpl just uses toString on the JsonObject. Really "JsonObject" could be renamed to "TextOrJsonObject". Though I haven't written a test for this scenario.

vkolomeyko replied:
I think it should work. `JsonObjectDeserializer` tries to parse content as `TreeNode` for non-Json content it will be a plain ValueNode with XML, CSV, etc. content.

conalsmith-r3 commented on applications/workers/workers-smoketest/src/smokeTest/kotlin/net/corda/applications/workers/smoketest/websocket/FlowStatusFeedSmokeTest.kt:
```diff
@@ -68,6 +68,8 @@ class FlowStatusFeedSmokeTest {
 
             eventually(Duration.ofSeconds(300)) {
```
I was seeing some issues with this test for some reason.

sprintengine commented on libs/http-rpc/http-rpc-server-impl/src/integrationTest/kotlin/net/corda/httprpc/server/impl/HttpRpcServerJsonObjectTest.kt:
```diff
@@ -0,0 +1,290 @@
+    private val expectedArrayResponse = """
+        {"id":"aaa123","obj":"[{\"message\":\"Hey Mars\",\"planetaryOnly\":\"true\",\"target\":\"C=GB, L=FOURTH, O=MARS, OU=PLANET\"},{\"message\":\"Hey Pluto\",\"planetaryOnly\":\"false\",\"target\":\"C=GB, L=FOURTH, O=PLUTO, OU=NON_PLANET\"}]"}
+    """.trimIndent()
+
```
I have created tests for the JSON being an object, a map and an array.

sprintengine commented on libs/http-rpc/http-rpc-test-common/src/main/kotlin/net/corda/httprpc/test/ObjectsInJsonEndpoint.kt:
```diff
@@ -0,0 +1,27 @@
+interface ObjectsInJsonEndpoint : RpcOps {
+
+    data class RequestWithJsonObject(val id: String, val obj: JsonObject)
+    data class ResponseWithJsonObject(val id: String, val obj: JsonObject)
```
These endpoints return the JsonObject, thus we can test the JSON marshalling too.

Atm we serialize the object and just take the escaped string.

lankydan reviewed (commented):
Keen to have this change, will let others comment more on the implementation, but the change in general makes sense to me.

vkolomeyko commented on libs/http-rpc/http-rpc-server-impl/src/main/kotlin/net/corda/httprpc/server/impl/internal/HttpRPCServerSerialization.kt:
```diff
@@ -26,3 +36,25 @@ internal object MemberX500NameSerializer : JsonSerializer<MemberX500Name>() {
+internal object JsonObjectDeserializer : JsonDeserializer<JsonObject>() {
+    override fun deserialize(p: JsonParser, ctxt: DeserializationContext): JsonObject {
+        val treeNode = p.readValueAsTree<TreeNode>()
+        val str = if (treeNode.isObject) {
```
Do we need this branch given that we have `else` with `treeNode.toString()` ?

sprintengine commented on libs/http-rpc/http-rpc-server-impl/src/main/kotlin/net/corda/httprpc/server/impl/internal/HttpRPCServerSerialization.kt:
```diff
@@ -26,3 +36,25 @@ internal object MemberX500NameSerializer : JsonSerializer<MemberX500Name>() {
+        generator.writeString(obj.toString())
+    }
+}
+
```
I can simplify the deserializer using `isContainerNode` instead of isObject and isArray. objects and arrays are container nodes

vkolomeyko commented on libs/http-rpc/http-rpc/src/main/kotlin/net/corda/httprpc/JsonObject.kt:
```diff
@@ -0,0 +1,32 @@
+ *
+ * Internally, implementations of [RpcOps] endpoints should use `toString()` on the [JsonObject] to get access to the object's JSON.
+ */
+interface JsonObject
```
I wonder if it would be neater to change this interface as follows:
```suggestion
interface JsonObject {
    val escapedJson: String
}
```
then we can use on `escapedJson` rather than rely on built-in `toString()` function.

sprintengine replied:
I didn't go with that approach because openApi generates a "JsonObject" with a property "escapedJson" and the purpose here is to allow real JSON rather than escaped JSON inside an object. The fact it escapes the Json in the deserializer is just internal. I don't think we need to expose this on the public API. 

The example payload in swagger if we did that:
```
{
  "clientRequestId": "string",
  "flowClassName": "string",
  "requestData": {
    "escapedJson": "string"
  }
}
```
`escapedJson` is a non-nullable string field and people would start using it like this:
```
{ 
    ...
    "requestData": {
        "escapedJson": "{\"command\":\"echo\", \"data\":{\"echo_value\": \"helloconal\"}}"
    }
}
```
When in reality we want them to use it like this:
```
{ 
    ...
    "requestData": "{\"command\":\"echo\", \"data\":{\"echo_value\": \"helloconal\"}}"
}
```
or
```
{ 
    ...
    "requestData": {
        "command":"echo", 
        "data":{
            "echo_value": "helloconal"
        }
    }
}
```
On the other hand, the current approach generates example request body with just empty object {} and this might be a bit confusing. Maybe we can improve this. Maybe we should introduce ignore annotation in a class so openApi ignores it
```
{
  "clientRequestId": "string",
  "flowClassName": "string",
  "requestData": {}
}
```

![image](https://user-images.githubusercontent.com/68279385/185452444-83fec4b4-e82c-4e5b-bff0-083b9647ac35.png)

lankydan replied:
Yeah, I was wondering what would happen to the example JSON in this case.

Maybe having the ability to specify example request bodies might be needed for these APIs? Not sure how else they can generate anything meaningful.

vkolomeyko replied:
I think there are two related aspects here:
1. The code in Kotlin that is easy to reason about;
2. Correct and meaningful representation of data structures in OpenAPI.

As for first point I think changing interface to:
```
interface JsonObject {
    val escapedJson: String
}
```
will make code look neater rather than we currently relying on `toString()` function. I totally appreciate that it has a negative impact on the OpenAPI we generate.

Which brings me to the second point, where we can add a custom `net.corda.httprpc.server.impl.apigen.processing.openapi.schema.builders.SchemaBuilder` for `JsonObject` where we will be able to control what gets written to OpenAPI. We can even provide a reasonable description and example here, I think.

sprintengine replied:
Yes that sounds reasonable, I'll apply those changes

vkolomeyko reviewed (commented):
Overall looks good. Few comments posted.

I am very happy to see that this has been introduced as a non-breaking change for existing invocations.

Probably for a separate PR, but I think we should change `UpdateConfigParameters.config` to `JsonObject` newly introduced type.

conal-smith committed:
CORE-6014 add "escapedJson" to JsonObject but hidden from OpenApi spec.
- composedSchema with anyOf for jsonObject indicating it accepts any json schema.
- added example and description to JsonObject
- simplify deserializer
- test nullability of JsonEndpoint

conal-smith commented:
Following latest changes, start flow API looks like this in open-api. Notice the example value shows a real json object in the requestData field.  (when we add some "example" to the annotations we can have per-API example JSON). We could change the example to be non-flow specific so it looks OK when we add JsonObject for config APIs.

![image](https://user-images.githubusercontent.com/68279385/186465183-85aeac8c-1871-42a5-8285-405705341f35.png)

Notice the description, example and nullable in the schema
![image](https://user-images.githubusercontent.com/68279385/186465265-5e3e8e4b-f99a-4985-ab79-4c370040248e.png)

Lets see if I can get a green build following some combined worker issues.

Adding JsonObject to config request objects can come in separate PR - https://r3-cev.atlassian.net/browse/CORE-6375

conal-smith committed:
CORE-6014 update swagger baseline following changes that remove separate schema for JsonObject and add anyOf fields.

vkolomeyko reviewed (approved):
LGTM
